grakn-graql/src/main/java/ai/grakn/graql/internal/query/ConceptBuilder.java

```diff
@@ -0,0 +1,334 @@
+import ai.grakn.concept.Concept;
+import ai.grakn.concept.ConceptId;
+import ai.grakn.concept.Label;
+import ai.grakn.concept.OntologyConcept;
+import ai.grakn.concept.Resource;
+import ai.grakn.concept.ResourceType;
+import ai.grakn.concept.Rule;
+import ai.grakn.concept.Thing;
+import ai.grakn.concept.Type;
+import ai.grakn.exception.GraqlQueryException;
+import ai.grakn.graql.Pattern;
+import ai.grakn.graql.Var;
+import ai.grakn.graql.internal.pattern.property.DataTypeProperty;
+import ai.grakn.graql.internal.pattern.property.IdProperty;
+import ai.grakn.graql.internal.pattern.property.IsaProperty;
+import ai.grakn.graql.internal.pattern.property.LabelProperty;
+import ai.grakn.graql.internal.pattern.property.SubProperty;
+import ai.grakn.graql.internal.pattern.property.ThenProperty;
+import ai.grakn.graql.internal.pattern.property.ValueProperty;
+import ai.grakn.graql.internal.pattern.property.WhenProperty;
+import ai.grakn.util.CommonUtil;
+import ai.grakn.util.Schema;
+
+import javax.annotation.Nullable;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.function.Function;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/**
+ * Class for building a {@link Concept}, by providing properties.
+ *
+ * <p>
+ *     A {@link ai.grakn.graql.admin.VarProperty} is responsible for inserting itself into the graph. However,
+ *     some properties can only operate in <i>combination</i>. For example, to create a {@link Resource} you need both
+ *     an {@link IsaProperty} and a {@link ValueProperty}.
+ * </p>
+ * <p>
+ *     Therefore, these properties do not create the {@link Concept} themselves.
+ *     instead they provide the necessary information to the {@link ConceptBuilder}, which will create the
+ *     {@link Concept} at a later time:
+ * </p>
+ * <pre>
+ *     // Executor:
+ *     ConceptBuilder builder = ConceptBuilder.of(executor, var);
+ *     // IsaProperty:
+ *     builder.isa(name);
+ *     // ValueProperty:
+ *     builder.value("Bob");
+ *     // Executor:
+ *     Concept concept = builder.build();
+ * </pre>
+ *
+ * @author Felix Chapman
+ */
+public class ConceptBuilder {
+
+    private final InsertQueryExecutor executor;
+
+    private final Var var;
+
+    /**
+     * A map of parameters that have been specified for this concept.
+     */
+    private final Map<BuilderParam<?>, Object> params = new HashMap<>();
+
+    /**
+     * A set of expected parameters, used when executing {@link #build()}.
+     * <p>
+     * Every time {@link #expect(BuilderParam)} or {@link #expectOrDefault(BuilderParam, Object)} is called, the argument is
+     * added to this set. After the concept is built, any parameter not in this set is considered "unexpected". If it
+     * is present in the field {@link #params}, then an error is thrown.
+     * </p>
+     */
+    private final Set<BuilderParam<?>> expectedParams = new HashSet<>();
```

Comment:
So this is like a list of "authorisedParams" ? So if something is added to `params` which is not in `expectedParams` you get an exception?